sql: simplify schema change backfill to not care for table versions

vivekmenezes opened the pull request:
The schema change backfill is applied as long as the table is not marked for deletion.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6361)

<!-- Reviewable:end -->

vivekmenezes committed:
sql: simplify schema change backfill to not care for table versions

The schema change backfill is applied as long as the table is not marked
for deletion.

andreimatei commented:
Should we now have a test that mutating a table while a schema change is running does not cause the schema change to have to be retried (whereas before it did)?

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSQUtwTKtOFh34V-Jz:-KGSQUtwTKtOFh34V-K-:-695340652) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L96)):*
s/return/returns
s/schemaChangeNoPending/schemaChangeNo*t*Pending

---

*[sql/backfill.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSQqLM-hFAVV2XYdF_:-KGSQqLM-hFAVV2XYdFa:677905754) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L100)):*
pErr

---

*[sql/backfill.go, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSXtTYcp7R3_GXpD1a:-KGSXtTYcp7R3_GXpD1b:-2109004750) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L101)):*
I think here we should convert whatever the "descriptor has been deleted" condition becomes into a `schemaChangeNotPendingError`. And then hopefully that allows us to remove the check for `DescriptorDeletedError` from the async execution path (timer at the bottom of `schema_changer.go`)

---

*[sql/backfill.go, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSR9LGBi5EZ_Ieh1rj:-KGSR9LGBi5EZ_Ieh1rk:-61482419) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L104)):*
this seems like an arbitrary check to be done by a `get` function. Let the caller do the get and call this `assertMutationPending`?

In fact, if the table has not been deleted, should the second condition ever fire? I think there should be no such thing as "another node has executed the schema change", apart from the deleted case, because of the schema change lease. That should be an unexpected error (and if we didn't have errors here, it'd be a panic.

So I think we should get rid of the `notPending` error and go back to `DescriptorDeletedError`.

---

*[sql/schema_changer.go, line 409 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSTVuXQGyrhnWFDfrq:-KGSTVuYBf9AOz9tTarB:238401789) ([raw file](https://github.com/cockroachdb/cockroach/blob/967711052744dd339ff9631f6e772157827aef40/sql/schema_changer.go#L409)):*
don't we want `IsDone` to not return an error if the table has been dropped?

---

*[sql/table.go, line 268 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSTtVi-gVaZhXog-YR:-KGSTtVi-gVaZhXog-YS:128521579) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/table.go#L268)):*
`DescriptorDeletedError` is something specific to schema changes, whereas this function is not. If you really want to return an error here, it should be `DescriptorNotFoundError`.
Or if this function is indeed specific to schema changes, move it away from this file and fuse it with the `getTableDesc` in `SchemaChanger`.

Anyway my preference is to leave this as a general function and have it return `nil`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-:-KGSVWCeoRGqMwhaUEqa:-882419847)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Added a test that deletes the table in the middle of a backfill

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSQUtwTKtOFh34V-Jz:-KGTOhk7QmRDqVPcvXmn:1587055311) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L96)):*
not relevant

---

*[sql/backfill.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSQqLM-hFAVV2XYdF_:-KGTOs21S8bY5L7QAsii:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L100)):*
done

---

*[sql/backfill.go, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSXtTYcp7R3_GXpD1a:-KGTPN9fAitDXqsEjvUF:463956080) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L101)):*
changed the async path to look for the NotFound error

---

*[sql/backfill.go, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSR9LGBi5EZ_Ieh1rj:-KGTPSRoieRLLCEHsw29:2032915057) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/backfill.go#L104)):*
Changed it to to only look for table.Deleted

---

*[sql/schema_changer.go, line 409 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSTVuXQGyrhnWFDfrq:-KGTPc2R6qERVSAXpUUb:-981184255) ([raw file](https://github.com/cockroachdb/cockroach/blob/967711052744dd339ff9631f6e772157827aef40/sql/schema_changer.go#L409)):*
I think it's okay for it to return an error if the table cannot be found.

---

*[sql/table.go, line 268 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGSTtVi-gVaZhXog-YR:-KGTPjetVjuss5UXQdIs:1446828475) ([raw file](https://github.com/cockroachdb/cockroach/blob/01e8ab6c366fa4e58810d4b736124b48189782c2/sql/table.go#L268)):*
It returns DescriptorNotFoundError now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-:-KGTOboX99VvWPruZQ6s:-3158519)*
<!-- Sent from Reviewable.io -->

andreimatei commented on sql/backfill.go:
```diff
@@ -282,7 +250,7 @@ func (sc *SchemaChanger) truncateAndBackfillColumns(
 
 			var currSentinel roachpb.Key
```
nit: s/currSentinel/curSentinel

andreimatei commented on sql/backfill.go:
```diff
@@ -89,46 +89,23 @@ func convertBackfillError(
-) (*TableDescriptor, *roachpb.Error) {
-	tableDesc, pErr := getTableDescFromID(txn, id)
+func (sc *SchemaChanger) getNonDeletedTableDesc(txn *client.Txn) (
+	*TableDescriptor, *roachpb.Error,
```
I believe the style says 
```
func (sc *SchemaChanger) getNonDeletedTableDesc(txn *client.Txn
) (*TableDescriptor, *roachpb.Error) {
```

andreimatei commented on sql/backfill.go:
```diff
@@ -89,46 +89,23 @@ func convertBackfillError(
-		return nil, descriptorChangedVersionError
+	if tableDesc.Deleted {
+		return nil, roachpb.NewError(
+			&roachpb.DescriptorNotFoundError{DescriptorId: uint32(sc.tableID)},
```
I think this error should be swallowed by all callers. So my usual nit applies: this function could return `nil`.

In fact, I wonder if we wouldn't need this function at all if we used leases (regular leases, not schema change leases). So the backfill would take and release a lease for for every span. In the common, non-contended case, this wouldn't cost anything. If the table has been marked deleted, the leasing system will know and not hand you a new lease (you might need a bit of care to make that you're releasing leases to the `LeaseManager` if going through the planner's leasing routines, since the planner also caches leases... I think, not sure).

andreimatei commented on sql/schema_changer_test.go:
```diff
@@ -667,4 +668,42 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+
+	backfillNotification = make(chan bool)
+	// Run the schema change in a separate goroutine.
+	var wg sync.WaitGroup
```
`wg` is not used

andreimatei commented on sql/schema_changer_test.go:
```diff
@@ -667,4 +668,42 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+	go func() {
+		// Start schema change that eventually runs a backfill.
+		_, err := sqlDB.Exec("CREATE UNIQUE INDEX bar ON t.test (v)")
+		if !testutils.IsError(err, "descriptor ID .* not found") {
```
but I thought the whole point is that the sql statement doesn't return an error. And this internal error looks like it definitely shouldn't be exposed.
So I'd swallow it in all the callers of `getDescriptorNotDeleted`. Which means you'd need a different mechanism to figure out that you've actually tested the "backfill encouters deleted table" code path. See my comment below about using KV "command filters" - you could use them to assert that some keys have never been backfilled, even though the `CREATE INDEX` statement finished.

andreimatei commented on sql/schema_changer_test.go:
```diff
@@ -667,4 +668,42 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+	// Wait for a short bit to ensure that the backfill has likely progressed
+	// and written some data, but not long enough that the backfill has
+	// completed.
+	time.Sleep(10 * time.Millisecond)
```
if you want to get cheeky you can replace this sleep with a KV command filter which watches for writes to the index. And this way you can also stall the backfill at an arbitrary point.

andreimatei commented on sql/schema_changer_test.go:
```diff
@@ -667,4 +668,42 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+	// Ensure that the table data has been deleted.
+	if kvs, err := kvDB.Scan(tablePrefix, tableEnd, 0); err != nil {
+		t.Fatal(err)
+	} else if e := 0; len(kvs) != e {
```
nit: just hardcode the 0

andreimatei commented:
LGTM but let's swallow that error

vivekmenezes commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGOVqPdDVpn1RwHbtg_-r2-93:-KGjB6IF0JEoY7DgubQu:878603312) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/backfill.go#L93)):*
removed this method.

---

*[sql/backfill.go, line 101 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGOVqPdDVpn1RwHbtg_-r2-101:-KGj27FruNXDSEDLA_DG:-385283452) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/backfill.go#L101)):*
I've taken this function out and explicitly check for the Deleted flag at the three call sites. I think it's cleaner now.

---

*[sql/backfill.go, line 251 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGOVqPdDVpn1RwHbtg_-r2-251:-KGj3LfE8ntReOIXC6J6:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/backfill.go#L251)):*
done

---

*[sql/schema_changer_test.go, line 681 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGTPuX2ryrl5RHCw6hC-r2-681:-KGj3gsWyMXnCpXZV4ER:-924673834) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/schema_changer_test.go#L681)):*
I use it now

---

*[sql/schema_changer_test.go, line 686 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGTPuX2ryrl5RHCw6hC-r2-686:-KGj84OcggYkqf4XWYMd:-1428213805) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/schema_changer_test.go#L686)):*
I now use the wg to wait until the backfill is done, and check that there is no garbage left over in the table after that.

---

*[sql/schema_changer_test.go, line 698 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGTPuX2ryrl5RHCw6hC-r2-698:-KGj8TaouHPAz5NSWKH0:-1735644172) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/schema_changer_test.go#L698)):*
true but this is a lot simpler and it is testing what we want in a random way.

---

*[sql/schema_changer_test.go, line 706 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361#-KGTPuX2ryrl5RHCw6hC-r2-706:-KGjAW5zRX68CSlXgDpM:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/b14b8960eb95af16d6dfbc8b9abf256d96bdf768/sql/schema_changer_test.go#L706)):*
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6361)*
<!-- Sent from Reviewable.io -->
